### DISCUSSION TITLE
More robust approach to data structure migration

### DISCUSSION THREAD

**Discussion by DillonJ** [spine-tools/SpineOpt.jl#870] [Ideas]
Hello all,

## TL/DR
 - We should consider specifying `upgrade=true` in the Run SpineOpt plugin tool and/or make `true` the default value for the keyword argument
 - The migration script should always automatically make a backup of the user's DB (I believe this is standard practice).

## Context

Recently I ran a simple test model to test a new installation of toolbox/SpineOpt. It failed - see the traceback below. The reason was that some of the new parameters associated with the new `unit_flow_capacity` constraint forumulation, introduced in recent updates, were not present. It took a little while to figure out as the traceback isn't very helpful, especially for a new user.

The solution was to call run_spineopt() with the keyword argument `upgrade = true` or re-import the latest template. 

The point of this is to highlight - that this might be a little troublesome for a new user or someone that isn't familliar with writing code. I think the main problem is that the Run SpineOpt plugin tool does not specify `upgrade = true` and the default value for this keyword argument is false. So all users with older DBs will encounter this show-stopping traceback unless they write some code.

When an old DB is run with `upgrade = true` the migration happens very gracefully.

Another aspect of the migration that we should consider is automatically backing up the old DB especially if we were to consider making `upgrade = true` the default, or at least specifying it in the Run SpineOpt plugin (and in the documentation/tutorials). There are a few reasons for this:
 - A user may wish to keep compatability with an older version of SpineOpt (e.g. perhaps they are running things on different machines that aren't all syncronised for some reason)
 - We might mess up with the migration script and/or there may be corner cases we didn't cover, so we don't want to irrecoverably corrupt a user's data

```julia
ERROR: LoadError: MethodError: no method matching zero(::Nothing)

Closest candidates are:
  zero(::Union{Type{P}, P}) where P<:Dates.Period
   @ Dates C:\Julia\Julia-1.9.3\share\julia\stdlib\v1.9\Dates\src\periods.jl:51
  zero(::JuMP.AbstractVariableRef)
   @ JuMP C:\Users\jodyd\.julia\packages\JuMP\D44Aq\src\variables.jl:348
  zero(::JuMP.GenericAffExpr)
   @ JuMP C:\Users\jodyd\.julia\packages\JuMP\D44Aq\src\aff_expr.jl:201
  ...

Stacktrace:
  [1] iszero(x::Nothing)
    @ Base .\number.jl:42
  [2] _diff_call(x::Int64, y::Nothing)
    @ SpineInterface C:\Workspace\Spine\spineinterface\src\base.jl:232
  [3] -(x::Int64, y::SpineInterface.Call)
    @ SpineInterface C:\Workspace\Spine\spineinterface\src\base.jl:228
  [4] _shutdown_margin(u::SpineInterface.Object, ng::SpineInterface.Object, d::SpineInterface.Object, s::SpineInterface.Object, t0::SpineInterface._StartRef, t::SpineInterface.TimeSlice, case::SpineInterface.Object, part::SpineInterface.Object)
    @ SpineOpt C:\Workspace\Spine\SpineOpt\src\constraints\constraint_unit_flow_capacity.jl:153
  [5] (::SpineOpt.var"#1079#1086"{SpineInterface.Object, SpineInterface.Object, SpineInterface.TimeSlice, SpineInterface.Object, SpineInterface.Object, SpineInterface._StartRef, Dict{NamedTuple{(:unit, :stochastic_scenario, :t), Tuple{SpineInterface.Object, SpineInterface.Object, SpineInterface.TimeSlice}}, JuMP.VariableRef}})(::NamedTuple{(:unit, :stochastic_scenario, :t), Tuple{SpineInterface.Object, SpineInterface.Object, SpineInterface.TimeSlice}})
    @ SpineOpt .\none:0
  [6] iterate
    @ .\generator.jl:47 [inlined]
  [7] first
    @ .\abstractarray.jl:465 [inlined]
  [8] expr_sum(iter::Base.Generator{Vector{NamedTuple{(:unit, :stochastic_scenario, :t), Tuple{SpineInterface.Object, SpineInterface.Object, SpineInterface.TimeSlice}}}, SpineOpt.var"#1079#1086"{SpineInterface.Object, SpineInterface.Object, SpineInterface.TimeSlice, SpineInterface.Object, SpineInterface.Object, SpineInterface._StartRef, Dict{NamedTuple{(:unit, :stochastic_scenario, :t), Tuple{SpineInterface.Object, SpineInterface.Object, SpineInterface.TimeSlice}}, JuMP.VariableRef}}}; init::Int64)
    @ SpineOpt C:\Workspace\Spine\SpineOpt\src\util\misc.jl:115
  [9] macro expansion
    @ C:\Users\jodyd\.julia\packages\MutableArithmetics\K9YPJ\src\rewrite.jl:321 [inlined]
 [10] macro expansion
    @ C:\Users\jodyd\.julia\packages\JuMP\D44Aq\src\macros.jl:717 [inlined]
 [11] #1075
    @ C:\Workspace\Spine\SpineOpt\src\constraints\constraint_unit_flow_capacity.jl:96 [inlined]
 [12] (::SpineOpt.var"#1075#1082"{JuMP.Model, SpineInterface._StartRef, Dict{NamedTuple{(:unit, :node, :stochastic_scenario, :t), Tuple{SpineInterface.Object, SpineInterface.Object, SpineInterface.Object, SpineInterface.TimeSlice}}, JuMP.VariableRef}, Dict{NamedTuple{(:unit, :node, :stochastic_scenario, :t), Tuple{SpineInterface.Object, SpineInterface.Object, SpineInterface.Object, SpineInterface.TimeSlice}}, JuMP.VariableRef}, Dict{NamedTuple{(:unit, :stochastic_scenario, :t), Tuple{SpineInterface.Object, SpineInterface.Object, SpineInterface.TimeSlice}}, JuMP.VariableRef}, Dict{NamedTuple{(:unit, :stochastic_scenario, :t), Tuple{SpineInterface.Object, SpineInterface.Object, SpineInterface.TimeSlice}}, JuMP.VariableRef}, Dict{NamedTuple{(:unit, :stochastic_scenario, :t), Tuple{SpineInterface.Object, SpineInterface.Object, SpineInterface.TimeSlice}}, JuMP.VariableRef}, Dict{NamedTuple{(:unit, :node, :direction, :stochastic_scenario, :t), Tuple{SpineInterface.Object, SpineInterface.Object, SpineInterface.Object, SpineInterface.Object, SpineInterface.TimeSlice}}, JuMP.VariableRef}})(::NamedTuple{(:unit, :node, :direction, :stochastic_path, :t, :t_next, :case, :part), Tuple{SpineInterface.Object, SpineInterface.Object, SpineInterface.Object, Vector{Any}, SpineInterface.TimeSlice, SpineInterface.TimeSlice, SpineInterface.Object, SpineInterface.Object}})
    @ SpineOpt .\none:0
 [13] iterate
    @ .\generator.jl:47 [inlined]
 [14] _all(f::Base.var"#372#374", itr::Base.Generator{Vector{NamedTuple{(:unit, :node, :direction, :stochastic_path, :t, :t_next, :case, :part), Tuple{SpineInterface.Object, SpineInterface.Object, SpineInterface.Object, Vector{Any}, SpineInterface.TimeSlice, SpineInterface.TimeSlice, SpineInterface.Object, SpineInterface.Object}}}, SpineOpt.var"#1075#1082"{JuMP.Model, SpineInterface._StartRef, Dict{NamedTuple{(:unit, :node, :stochastic_scenario, :t), Tuple{SpineInterface.Object, SpineInterface.Object, SpineInterface.Object, SpineInterface.TimeSlice}}, JuMP.VariableRef}, Dict{NamedTuple{(:unit, :node, :stochastic_scenario, :t), Tuple{SpineInterface.Object, SpineInterface.Object, SpineInterface.Object, SpineInterface.TimeSlice}}, JuMP.VariableRef}, Dict{NamedTuple{(:unit, :stochastic_scenario, :t), Tuple{SpineInterface.Object, SpineInterface.Object, SpineInterface.TimeSlice}}, JuMP.VariableRef}, Dict{NamedTuple{(:unit, :stochastic_scenario, :t), Tuple{SpineInterface.Object, SpineInterface.Object, SpineInterface.TimeSlice}}, JuMP.VariableRef}, Dict{NamedTuple{(:unit, :stochastic_scenario, :t), Tuple{SpineInterface.Object, SpineInterface.Object, SpineInterface.TimeSlice}}, JuMP.VariableRef}, Dict{NamedTuple{(:unit, :node, :direction, :stochastic_scenario, :t), Tuple{SpineInterface.Object, SpineInterface.Object, SpineInterface.Object, SpineInterface.Object, SpineInterface.TimeSlice}}, JuMP.VariableRef}}}, #unused#::Colon)
    @ Base .\reduce.jl:1282
 [15] all
    @ .\reduce.jl:1278 [inlined]
 [16] Dict(kv::Base.Generator{Vector{NamedTuple{(:unit, :node, :direction, :stochastic_path, :t, :t_next, :case, :part), Tuple{SpineInterface.Object, SpineInterface.Object, SpineInterface.Object, Vector{Any}, SpineInterface.TimeSlice, SpineInterface.TimeSlice, SpineInterface.Object, SpineInterface.Object}}}, SpineOpt.var"#1075#1082"{JuMP.Model, SpineInterface._StartRef, Dict{NamedTuple{(:unit, :node, :stochastic_scenario, :t), Tuple{SpineInterface.Object, SpineInterface.Object, SpineInterface.Object, SpineInterface.TimeSlice}}, JuMP.VariableRef}, Dict{NamedTuple{(:unit, :node, :stochastic_scenario, :t), Tuple{SpineInterface.Object, SpineInterface.Object, SpineInterface.Object, SpineInterface.TimeSlice}}, JuMP.VariableRef}, Dict{NamedTuple{(:unit, :stochastic_scenario, :t), Tuple{SpineInterface.Object, SpineInterface.Object, SpineInterface.TimeSlice}}, JuMP.VariableRef}, Dict{NamedTuple{(:unit, :stochastic_scenario, :t), Tuple{SpineInterface.Object, SpineInterface.Object, SpineInterface.TimeSlice}}, JuMP.VariableRef}, Dict{NamedTuple{(:unit, :stochastic_scenario, :t), Tuple{SpineInterface.Object, SpineInterface.Object, SpineInterface.TimeSlice}}, JuMP.VariableRef}, Dict{NamedTuple{(:unit, :node, :direction, :stochastic_scenario, :t), Tuple{SpineInterface.Object, SpineInterface.Object, SpineInterface.Object, SpineInterface.Object, SpineInterface.TimeSlice}}, JuMP.VariableRef}}})
    @ Base .\dict.jl:111
 [17] add_constraint_unit_flow_capacity!(m::JuMP.Model)
    @ SpineOpt C:\Workspace\Spine\SpineOpt\src\constraints\constraint_unit_flow_capacity.jl:95
 [18] macro expansion
    @ .\timing.jl:273 [inlined]
 [19] macro expansion
    @ C:\Workspace\Spine\SpineOpt\src\util\misc.jl:51 [inlined]
 [20] _add_constraints!(m::JuMP.Model; add_constraints::SpineOpt.var"#42#52", log_level::Int64)
    @ SpineOpt C:\Workspace\Spine\SpineOpt\src\run_spineopt_standard.jl:209
 [21] macro expansion
    @ .\timing.jl:273 [inlined]
 [22] macro expansion
    @ C:\Workspace\Spine\SpineOpt\src\util\misc.jl:51 [inlined]
 [23] init_model!(m::JuMP.Model; add_user_variables::Function, add_constraints::Function, log_level::Int64)
    @ SpineOpt C:\Workspace\Spine\SpineOpt\src\run_spineopt_standard.jl:79
 [24] init_model!
    @ C:\Workspace\Spine\SpineOpt\src\run_spineopt_standard.jl:75 [inlined]
 [25] rerun_spineopt_standard!(m::JuMP.Model, url_out::String; add_user_variables::Function, add_constraints::Function, log_level::Int64, optimize::Bool, update_names::Bool, alternative::String, write_as_roll::Int64, resume_file_path::Nothing, run_kernel::typeof(run_spineopt_kernel!))
    @ SpineOpt C:\Workspace\Spine\SpineOpt\src\run_spineopt_standard.jl:37
 [26] invokelatest(::Any, ::Any, ::Vararg{Any}; kwargs::Base.Pairs{Symbol, Any, NTuple{9, Symbol}, NamedTuple{(:add_user_variables, :add_constraints, :log_level, :optimize, :update_names, :alternative, :write_as_roll, :resume_file_path, :run_kernel), Tuple{SpineOpt.var"#41#51", SpineOpt.var"#42#52", Int64, Bool, Bool, String, Int64, Nothing, typeof(run_spineopt_kernel!)}}})
    @ Base .\essentials.jl:821
 [27] rerun_spineopt(url_out::String; mip_solver::Nothing, lp_solver::Nothing, add_user_variables::Function, add_constraints::Function, log_level::Int64, optimize::Bool, update_names::Bool, alternative::String, write_as_roll::Int64, resume_file_path::Nothing, use_direct_model::Bool, run_kernel::typeof(run_spineopt_kernel!))
    @ SpineOpt C:\Workspace\Spine\SpineOpt\src\run_spineopt.jl:258
 [28] rerun_spineopt
    @ C:\Workspace\Spine\SpineOpt\src\run_spineopt.jl:235 [inlined]
 [29] _run_spineopt(url_in::String, url_out::String; upgrade::Bool, log_level::Int64, filters::Dict{String, String}, templates::Tuple{}, kwargs::Base.Pairs{Symbol, Any, NTuple{11, Symbol}, NamedTuple{(:mip_solver, :lp_solver, :add_user_variables, :add_constraints, :optimize, :update_names, :alternative, :write_as_roll, :use_direct_model, :resume_file_path, :run_kernel), Tuple{Nothing, Nothing, SpineOpt.var"#41#51", SpineOpt.var"#42#52", Bool, Bool, String, Int64, Bool, Nothing, typeof(run_spineopt_kernel!)}}})
    @ SpineOpt C:\Workspace\Spine\SpineOpt\src\run_spineopt.jl:178
 [30] _run_spineopt
    @ C:\Workspace\Spine\SpineOpt\src\run_spineopt.jl:174 [inlined]
 [31] #run_spineopt#34
    @ C:\Workspace\Spine\SpineOpt\src\run_spineopt.jl:101 [inlined]
 [32] run_spineopt(url_in::String, url_out::String)
    @ SpineOpt C:\Workspace\Spine\SpineOpt\src\run_spineopt.jl:80
 [33] top-level scope
    @ C:\Users\jodyd\.spinetoolbox\work\run_spineopt__5d7c061a315f4a5a8b76d69c10b24e7e__toolbox\run_spineopt.jl:3
 [34] include(fname::String)
    @ Base.MainInclude .\client.jl:478
 [35] top-level scope
    @ none:1
in expression starting at C:\Users\jodyd\.spinetoolbox\work\run_spineopt__5d7c061a315f4a5a8b76d69c10b24e7e__toolbox\run_spineopt.jl:3

caused by: MethodError: no method matching zero(::Nothing)

Closest candidates are:
  zero(::Union{Type{P}, P}) where P<:Dates.Period
   @ Dates C:\Julia\Julia-1.9.3\share\julia\stdlib\v1.9\Dates\src\periods.jl:51
  zero(::JuMP.AbstractVariableRef)
   @ JuMP C:\Users\jodyd\.julia\packages\JuMP\D44Aq\src\variables.jl:348
  zero(::JuMP.GenericAffExpr)
   @ JuMP C:\Users\jodyd\.julia\packages\JuMP\D44Aq\src\aff_expr.jl:201
  ...

Stacktrace:
  [1] iszero(x::Nothing)
    @ Base .\number.jl:42
  [2] _diff_call(x::Int64, y::Nothing)
    @ SpineInterface C:\Workspace\Spine\spineinterface\src\base.jl:232
  [3] -(x::Int64, y::SpineInterface.Call)
    @ SpineInterface C:\Workspace\Spine\spineinterface\src\base.jl:228
  [4] _shutdown_margin(u::SpineInterface.Object, ng::SpineInterface.Object, d::SpineInterface.Object, s::SpineInterface.Object, t0::SpineInterface._StartRef, t::SpineInterface.TimeSlice, case::SpineInterface.Object, part::SpineInterface.Object)
    @ SpineOpt C:\Workspace\Spine\SpineOpt\src\constraints\constraint_unit_flow_capacity.jl:153
  [5] (::SpineOpt.var"#1079#1086"{SpineInterface.Object, SpineInterface.Object, SpineInterface.TimeSlice, SpineInterface.Object, SpineInterface.Object, SpineInterface._StartRef, Dict{NamedTuple{(:unit, :stochastic_scenario, :t), Tuple{SpineInterface.Object, SpineInterface.Object, SpineInterface.TimeSlice}}, JuMP.VariableRef}})(::NamedTuple{(:unit, :stochastic_scenario, :t), Tuple{SpineInterface.Object, SpineInterface.Object, SpineInterface.TimeSlice}})
    @ SpineOpt .\none:0
  [6] iterate
    @ .\generator.jl:47 [inlined]
  [7] first
    @ .\abstractarray.jl:465 [inlined]
  [8] expr_sum(iter::Base.Generator{Vector{NamedTuple{(:unit, :stochastic_scenario, :t), Tuple{SpineInterface.Object, SpineInterface.Object, SpineInterface.TimeSlice}}}, SpineOpt.var"#1079#1086"{SpineInterface.Object, SpineInterface.Object, SpineInterface.TimeSlice, SpineInterface.Object, SpineInterface.Object, SpineInterface._StartRef, Dict{NamedTuple{(:unit, :stochastic_scenario, :t), Tuple{SpineInterface.Object, SpineInterface.Object, SpineInterface.TimeSlice}}, JuMP.VariableRef}}}; init::Int64)
    @ SpineOpt C:\Workspace\Spine\SpineOpt\src\util\misc.jl:115
  [9] macro expansion
    @ C:\Users\jodyd\.julia\packages\MutableArithmetics\K9YPJ\src\rewrite.jl:321 [inlined]
 [10] macro expansion
    @ C:\Users\jodyd\.julia\packages\JuMP\D44Aq\src\macros.jl:717 [inlined]
 [11] #1075
    @ C:\Workspace\Spine\SpineOpt\src\constraints\constraint_unit_flow_capacity.jl:96 [inlined]
 [12] (::SpineOpt.var"#1075#1082"{JuMP.Model, SpineInterface._StartRef, Dict{NamedTuple{(:unit, :node, :stochastic_scenario, :t), Tuple{SpineInterface.Object, SpineInterface.Object, SpineInterface.Object, SpineInterface.TimeSlice}}, JuMP.VariableRef}, Dict{NamedTuple{(:unit, :node, :stochastic_scenario, :t), Tuple{SpineInterface.Object, SpineInterface.Object, SpineInterface.Object, SpineInterface.TimeSlice}}, JuMP.VariableRef}, Dict{NamedTuple{(:unit, :stochastic_scenario, :t), Tuple{SpineInterface.Object, SpineInterface.Object, SpineInterface.TimeSlice}}, JuMP.VariableRef}, Dict{NamedTuple{(:unit, :stochastic_scenario, :t), Tuple{SpineInterface.Object, SpineInterface.Object, SpineInterface.TimeSlice}}, JuMP.VariableRef}, Dict{NamedTuple{(:unit, :stochastic_scenario, :t), Tuple{SpineInterface.Object, SpineInterface.Object, SpineInterface.TimeSlice}}, JuMP.VariableRef}, Dict{NamedTuple{(:unit, :node, :direction, :stochastic_scenario, :t), Tuple{SpineInterface.Object, SpineInterface.Object, SpineInterface.Object, SpineInterface.Object, SpineInterface.TimeSlice}}, JuMP.VariableRef}})(::NamedTuple{(:unit, :node, :direction, :stochastic_path, :t, :t_next, :case, :part), Tuple{SpineInterface.Object, SpineInterface.Object, SpineInterface.Object, Vector{Any}, SpineInterface.TimeSlice, SpineInterface.TimeSlice, SpineInterface.Object, SpineInterface.Object}})
    @ SpineOpt .\none:0
 [13] iterate
    @ .\generator.jl:47 [inlined]
 [14] grow_to!(dest::Dict{Any, Any}, itr::Base.Generator{Vector{NamedTuple{(:unit, :node, :direction, :stochastic_path, :t, :t_next, :case, :part), Tuple{SpineInterface.Object, SpineInterface.Object, SpineInterface.Object, Vector{Any}, SpineInterface.TimeSlice, SpineInterface.TimeSlice, SpineInterface.Object, SpineInterface.Object}}}, SpineOpt.var"#1075#1082"{JuMP.Model, SpineInterface._StartRef, Dict{NamedTuple{(:unit, :node, :stochastic_scenario, :t), Tuple{SpineInterface.Object, SpineInterface.Object, SpineInterface.Object, SpineInterface.TimeSlice}}, JuMP.VariableRef}, Dict{NamedTuple{(:unit, :node, :stochastic_scenario, :t), Tuple{SpineInterface.Object, SpineInterface.Object, SpineInterface.Object, SpineInterface.TimeSlice}}, JuMP.VariableRef}, Dict{NamedTuple{(:unit, :stochastic_scenario, :t), Tuple{SpineInterface.Object, SpineInterface.Object, SpineInterface.TimeSlice}}, JuMP.VariableRef}, Dict{NamedTuple{(:unit, :stochastic_scenario, :t), Tuple{SpineInterface.Object, SpineInterface.Object, SpineInterface.TimeSlice}}, JuMP.VariableRef}, Dict{NamedTuple{(:unit, :stochastic_scenario, :t), Tuple{SpineInterface.Object, SpineInterface.Object, SpineInterface.TimeSlice}}, JuMP.VariableRef}, Dict{NamedTuple{(:unit, :node, :direction, :stochastic_scenario, :t), Tuple{SpineInterface.Object, SpineInterface.Object, SpineInterface.Object, SpineInterface.Object, SpineInterface.TimeSlice}}, JuMP.VariableRef}}})
    @ Base .\dict.jl:120
 [15] dict_with_eltype
    @ .\abstractdict.jl:590 [inlined]
 [16] Dict(kv::Base.Generator{Vector{NamedTuple{(:unit, :node, :direction, :stochastic_path, :t, :t_next, :case, :part), Tuple{SpineInterface.Object, SpineInterface.Object, SpineInterface.Object, Vector{Any}, SpineInterface.TimeSlice, SpineInterface.TimeSlice, SpineInterface.Object, SpineInterface.Object}}}, SpineOpt.var"#1075#1082"{JuMP.Model, SpineInterface._StartRef, Dict{NamedTuple{(:unit, :node, :stochastic_scenario, :t), Tuple{SpineInterface.Object, SpineInterface.Object, SpineInterface.Object, SpineInterface.TimeSlice}}, JuMP.VariableRef}, Dict{NamedTuple{(:unit, :node, :stochastic_scenario, :t), Tuple{SpineInterface.Object, SpineInterface.Object, SpineInterface.Object, SpineInterface.TimeSlice}}, JuMP.VariableRef}, Dict{NamedTuple{(:unit, :stochastic_scenario, :t), Tuple{SpineInterface.Object, SpineInterface.Object, SpineInterface.TimeSlice}}, JuMP.VariableRef}, Dict{NamedTuple{(:unit, :stochastic_scenario, :t), Tuple{SpineInterface.Object, SpineInterface.Object, SpineInterface.TimeSlice}}, JuMP.VariableRef}, Dict{NamedTuple{(:unit, :stochastic_scenario, :t), Tuple{SpineInterface.Object, SpineInterface.Object, SpineInterface.TimeSlice}}, JuMP.VariableRef}, Dict{NamedTuple{(:unit, :node, :direction, :stochastic_scenario, :t), Tuple{SpineInterface.Object, SpineInterface.Object, SpineInterface.Object, SpineInterface.Object, SpineInterface.TimeSlice}}, JuMP.VariableRef}}})
    @ Base .\dict.jl:109
 [17] add_constraint_unit_flow_capacity!(m::JuMP.Model)
    @ SpineOpt C:\Workspace\Spine\SpineOpt\src\constraints\constraint_unit_flow_capacity.jl:95
 [18] macro expansion
    @ .\timing.jl:273 [inlined]
 [19] macro expansion
    @ C:\Workspace\Spine\SpineOpt\src\util\misc.jl:51 [inlined]
 [20] _add_constraints!(m::JuMP.Model; add_constraints::SpineOpt.var"#42#52", log_level::Int64)
    @ SpineOpt C:\Workspace\Spine\SpineOpt\src\run_spineopt_standard.jl:209
 [21] macro expansion
    @ .\timing.jl:273 [inlined]
 [22] macro expansion
    @ C:\Workspace\Spine\SpineOpt\src\util\misc.jl:51 [inlined]
 [23] init_model!(m::JuMP.Model; add_user_variables::Function, add_constraints::Function, log_level::Int64)
    @ SpineOpt C:\Workspace\Spine\SpineOpt\src\run_spineopt_standard.jl:79
 [24] init_model!
    @ C:\Workspace\Spine\SpineOpt\src\run_spineopt_standard.jl:75 [inlined]
 [25] rerun_spineopt_standard!(m::JuMP.Model, url_out::String; add_user_variables::Function, add_constraints::Function, log_level::Int64, optimize::Bool, update_names::Bool, alternative::String, write_as_roll::Int64, resume_file_path::Nothing, run_kernel::typeof(run_spineopt_kernel!))
    @ SpineOpt C:\Workspace\Spine\SpineOpt\src\run_spineopt_standard.jl:37
 [26] invokelatest(::Any, ::Any, ::Vararg{Any}; kwargs::Base.Pairs{Symbol, Any, NTuple{9, Symbol}, NamedTuple{(:add_user_variables, :add_constraints, :log_level, :optimize, :update_names, :alternative, :write_as_roll, :resume_file_path, :run_kernel), Tuple{SpineOpt.var"#41#51", SpineOpt.var"#42#52", Int64, Bool, Bool, String, Int64, Nothing, typeof(run_spineopt_kernel!)}}})
    @ Base .\essentials.jl:821
 [27] rerun_spineopt(url_out::String; mip_solver::Nothing, lp_solver::Nothing, add_user_variables::Function, add_constraints::Function, log_level::Int64, optimize::Bool, update_names::Bool, alternative::String, write_as_roll::Int64, resume_file_path::Nothing, use_direct_model::Bool, run_kernel::typeof(run_spineopt_kernel!))
    @ SpineOpt C:\Workspace\Spine\SpineOpt\src\run_spineopt.jl:258
 [28] rerun_spineopt
    @ C:\Workspace\Spine\SpineOpt\src\run_spineopt.jl:235 [inlined]
 [29] _run_spineopt(url_in::String, url_out::String; upgrade::Bool, log_level::Int64, filters::Dict{String, String}, templates::Tuple{}, kwargs::Base.Pairs{Symbol, Any, NTuple{11, Symbol}, NamedTuple{(:mip_solver, :lp_solver, :add_user_variables, :add_constraints, :optimize, :update_names, :alternative, :write_as_roll, :use_direct_model, :resume_file_path, :run_kernel), Tuple{Nothing, Nothing, SpineOpt.var"#41#51", SpineOpt.var"#42#52", Bool, Bool, String, Int64, Bool, Nothing, typeof(run_spineopt_kernel!)}}})
    @ SpineOpt C:\Workspace\Spine\SpineOpt\src\run_spineopt.jl:178
 [30] _run_spineopt
    @ C:\Workspace\Spine\SpineOpt\src\run_spineopt.jl:174 [inlined]
 [31] #run_spineopt#34
    @ C:\Workspace\Spine\SpineOpt\src\run_spineopt.jl:101 [inlined]
 [32] run_spineopt(url_in::String, url_out::String)
    @ SpineOpt C:\Workspace\Spine\SpineOpt\src\run_spineopt.jl:80
 [33] top-level scope
    @ C:\Users\jodyd\.spinetoolbox\work\run_spineopt__5d7c061a315f4a5a8b76d69c10b24e7e__toolbox\run_spineopt.jl:3
 [34] include(fname::String)
    @ Base.MainInclude .\client.jl:478
 [35] top-level scope
    @ none:1
```
